In [1]:
# Import the modules
import numpy as np
import pandas as pd
from pathlib import Path
from sklearn.metrics import confusion_matrix, classification_report

---

## Split the Data into Training and Testing Sets

### Step 1: Read the `lending_data.csv` data from the `Resources` folder into a Pandas DataFrame.

In [2]:
# Read the CSV file from the Resources folder into a Pandas DataFrame
df = pd.read_csv(
    Path('Resources/lending_data.csv')   
)

# Review the DataFrame
df.head()

loan_size  interest_rate  borrower_income  debt_to_income  num_of_accounts  \
0    10700.0          7.672            52800        0.431818                5   
1     8400.0          6.692            43600        0.311927                3   
2     9000.0          6.963            46100        0.349241                3   
3    10700.0          7.664            52700        0.430740                5   
4    10800.0          7.698            53000        0.433962                5   

   derogatory_marks  total_debt  loan_status  
0                 1       22800            0  
1                 0       13600            0  
2                 0       16100            0  
3                 1       22700            0  
4                 1       23000            0

### Step 2: Create the labels set (`y`)  from the “loan_status” column, and then create the features (`X`) DataFrame from the remaining columns.

In [3]:
# Separate the data into labels and features

# Separate the y variable, the labels
y = df['loan_status']

# Separate the X variable, the features
X = df.drop(columns='loan_status')

In [4]:
# Review the y variable Series
print(y)

0        0
1        0
2        0
3        0
4        0
        ..
77531    1
77532    1
77533    1
77534    1
77535    1
Name: loan_status, Length: 77536, dtype: int64


In [5]:
# Review the X variable DataFrame
X.head()

loan_size  interest_rate  borrower_income  debt_to_income  num_of_accounts  \
0    10700.0          7.672            52800        0.431818                5   
1     8400.0          6.692            43600        0.311927                3   
2     9000.0          6.963            46100        0.349241                3   
3    10700.0          7.664            52700        0.430740                5   
4    10800.0          7.698            53000        0.433962                5   

   derogatory_marks  total_debt  
0                 1       22800  
1                 0       13600  
2                 0       16100  
3                 1       22700  
4                 1       23000

### Step 3: Split the data into training and testing datasets by using `train_test_split`.

In [6]:
# Import the train_test_learn module
from sklearn.model_selection import train_test_split

# Split the data using train_test_split
# Assign a random_state of 1 to the function
X_train, X_test, y_train, y_test = train_test_split(X, y, random_state=1)

---

## Create a Logistic Regression Model with the Original Data

###  Step 1: Fit a logistic regression model by using the training data (`X_train` and `y_train`).

In [7]:
# Import the LogisticRegression module from SKLearn
from sklearn.linear_model import LogisticRegression

# Instantiate the Logistic Regression model
# Assign a random_state parameter of 1 to the model
classifier = LogisticRegression(solver='lbfgs', random_state=1)
classifier

# Fit the model using training data
classifier.fit(X_train, y_train)

LogisticRegression(random_state=1)

### Step 2: Save the predictions on the testing data labels by using the testing feature data (`X_test`) and the fitted model.

In [8]:
# Make a prediction using the testing data
predictions = classifier.predict(X_test)
pd.DataFrame({"Prediction": predictions, "Actual": y_test})

Prediction  Actual
60914           0       0
36843           0       0
1966            0       0
70137           0       0
27237           0       0
...           ...     ...
45639           0       0
11301           0       0
51614           0       0
4598            0       0
2793            0       0

[19384 rows x 2 columns]

### Step 3: Evaluate the model’s performance by doing the following:

* Generate a confusion matrix.

* Print the classification report.

In [9]:
# Generate a confusion matrix for the model
from sklearn.metrics import confusion_matrix
confusion_matrix(y_test, predictions)

array([[18655,   110],
       [   36,   583]], dtype=int64)

In [10]:
# Print the classification report for the model
from sklearn.metrics import classification_report
testing_report = classification_report(y_test, predictions)
print(testing_report)

              precision    recall  f1-score   support

           0       1.00      0.99      1.00     18765
           1       0.84      0.94      0.89       619

    accuracy                           0.99     19384
   macro avg       0.92      0.97      0.94     19384
weighted avg       0.99      0.99      0.99     19384



### Step 4: Answer the following question.

**Question:** How well does the logistic regression model predict both the `0` (healthy loan) and `1` (high-risk loan) labels?

**Answer:** The model performs exceptionally well in predicting Class 0 (healthy loans) with near-perfect precision, recall, and F1-scores, indicating accurate identification and minimal misclassification. For Class 1 (high-risk loans), performance is strong but slightly lower, with a precision of 0.84 and recall of 0.94, capturing most high-risk loans but with some false positives. Overall accuracy is high (0.99), driven by correct predictions of healthy loans. However, there is room for improvement in refining high-risk loan identification to achieve balanced class performance.

---

Code to Train and Evaluate Random Forest, SVM, KNN, and Decision Tree Models

In [12]:
# Import necessary libraries
from sklearn.ensemble import RandomForestClassifier
from sklearn.svm import SVC
from sklearn.neighbors import KNeighborsClassifier
from sklearn.tree import DecisionTreeClassifier
from sklearn.metrics import classification_report, confusion_matrix



# Random Forest Model
rf_model = RandomForestClassifier(random_state=1)
rf_model.fit(X_train, y_train)
rf_predictions = rf_model.predict(X_test)
print("Random Forest Model Performance:")
print(confusion_matrix(y_test, rf_predictions))
print(classification_report(y_test, rf_predictions))

# Support Vector Machine (SVM) Model
svm_model = SVC(random_state=1)
svm_model.fit(X_train, y_train)
svm_predictions = svm_model.predict(X_test)
print("SVM Model Performance:")
print(confusion_matrix(y_test, svm_predictions))
print(classification_report(y_test, svm_predictions))

# K-Nearest Neighbors (KNN) Model
knn_model = KNeighborsClassifier()
knn_model.fit(X_train, y_train)
knn_predictions = knn_model.predict(X_test)
print("KNN Model Performance:")
print(confusion_matrix(y_test, knn_predictions))
print(classification_report(y_test, knn_predictions))

# Decision Tree Model
dt_model = DecisionTreeClassifier(random_state=1)
dt_model.fit(X_train, y_train)
dt_predictions = dt_model.predict(X_test)
print("Decision Tree Model Performance:")
print(confusion_matrix(y_test, dt_predictions))
print(classification_report(y_test, dt_predictions))

Random Forest Model Performance:
[[18666    99]
 [   66   553]]
              precision    recall  f1-score   support

           0       1.00      0.99      1.00     18765
           1       0.85      0.89      0.87       619

    accuracy                           0.99     19384
   macro avg       0.92      0.94      0.93     19384
weighted avg       0.99      0.99      0.99     19384

SVM Model Performance:
[[18651   114]
 [    6   613]]
              precision    recall  f1-score   support

           0       1.00      0.99      1.00     18765
           1       0.84      0.99      0.91       619

    accuracy                           0.99     19384
   macro avg       0.92      0.99      0.95     19384
weighted avg       0.99      0.99      0.99     19384

KNN Model Performance:
[[18651   114]
 [   16   603]]
              precision    recall  f1-score   support

           0       1.00      0.99      1.00     18765
           1       0.84      0.97      0.90       619

    accura

Summary: Refined Analysis and Recommendation
Best Performing Model: The SVM model demonstrates slightly better performance for high-risk loans (Class 1), with the highest recall (0.99) and the best F1-score (0.91). This makes it particularly effective for capturing a high proportion of high-risk loans, thereby minimizing false negatives. Moreover, it maintains strong performance for identifying healthy loans (Class 0).

Context-Specific Considerations:

Minimizing False Positives: If prioritizing high precision for high-risk loans (i.e., reducing the number of healthy loans misclassified as high-risk) is critical, Random Forest or KNN could be more suitable, as they offer strong performance while slightly sacrificing recall.
Overall Performance Balance: For high overall accuracy and a focus on minimizing false negatives, SVM remains a strong choice due to its balanced precision and recall across both classes.
Recommendation: Since accurately identifying high-risk loans with a high recall is crucial for credit risk management (to avoid missing potential defaulters), the SVM model is recommended. It provides a well-rounded approach by capturing nearly all high-risk loans while delivering excellent accuracy for healthy loans, making it an optimal balance of precision and recall for the task at hand.